### PR TITLE
Use the correct types for generated ids

### PR DESCRIPTION
**Pull request by lcobucci**
This merges #5935 and #6020 in order to solve #5684.

### CODE REVIEW

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 1]
```diff
@@ -890,7 +887,14 @@ private function persistNew($class, $entity)
             $idValue = $idGen->generate($this->em, $entity);
 
             if ( ! $idGen instanceof \Doctrine\ORM\Id\AssignedGenerator) {
-                $idValue = array($class->identifier[0] => $idValue);
+                $idField = $class->identifier[0];
```
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: I just digged into this conversion in depth, and it seems wrong. The `AssignedGenerator` already uses a PHP value internally, and never produces an SQL value. This code can therefore be dropped, and test code needs to be re-validated

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 1 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: Nvm, applies to non-assigned generators -.- Misread the `!` in front of it.

Comment (lcobucci) [inline lib/Doctrine/ORM/UnitOfWork.php]: But then it's a bit confusing... should this conversion be generators' responsibility?

Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: Could be, but can't refactor it now, since generators have very little access to external scope.

**Comment by Ocramius** [comment]
@lcobucci merged and closed all related issues, thanks a lot!

Note: I added a few additional commits to just extract the common code.

**Pull request merged**